# Getting our data ready for ML via sklearn

Yeah, I know, it feels kind of difficult... <br>
![](https://media.giphy.com/media/1eBQ8eo1AYyGc/giphy.gif) ![](https://media.giphy.com/media/hSUAf0zd1u56U/giphy.gif)

See? I don't always use memes :)
Best wishes to you!

## Standard imports

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### Our objectives 
#### 1. Split data into features (x usually) and labels (y usually)
#### 2. Filling (also called "Imputing") or disregarding missing values
#### 3. Converting non-numerical values to numerical values (a.k.a Feature encoding)

In [2]:
# Let's begin with the heart disease dataset
hd = pd.read_csv("https://raw.githubusercontent.com/ineelhere/Machine-Learning-and-Data-Science/master/scikit-learn/heart-disease.csv")
hd.head(3)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1

In [3]:
#length/size of the dataset
len(hd)

303

In [4]:
# get the feature (x) ready = all columns except target
x = hd.drop("target", axis=1)
x.head(3)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2

In [5]:
# get the labels (y) ready = target column
y = hd.target
y.head(3)

0    1
1    1
2    1
Name: target, dtype: int64

# Always Remember! <br>
## NEVER  EVALUATE  or  TEST your models on DATA that it is LEARNT FROM -- that's why we split it into training and test sets.
(because if you do, it is like cheating in an examination)  <br>
![](https://media.giphy.com/media/A4CLbWb9o5rj2/giphy.gif)

### 1. Splitting the data into test and training sets
Imagine it this way
* Test data = Final exam
* Training data = Mock exam (So we use the train_test_split first!)

In [6]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.2) 
# test_size=0.2 means that we want our test dataset to be 20% of the overall data

In [7]:
#check the shapes of the new matrices just created. 
x_train.shape, x_test.shape, y_train.shape, y_test.shape

((242, 13), (61, 13), (242,), (61,))

Just notice,  <br>
242 = 80% of 303 <br>
61 = 20% of 303 <br>
So, training set is 80% of overall data and test set is 20% of the overall data  <br>
(I have mentioned this above too!)

### 2. Filling (also called "Imputing") or disregarding missing values
### Make sure data is in numerial format/values  - if not, make them!

In [8]:
#let us use a new dataset now. So first we will get the dataset
cse = pd.read_csv("https://raw.githubusercontent.com/ineelhere/Machine-Learning-and-Data-Science/master/scikit-learn/car-sales-extended.csv")
cse.head()
#ummm... I mean car-sales-extended by cse. (I'm lazy!)

Make Colour  Odometer (KM)  Doors  Price
0   Honda  White          35431      4  15323
1     BMW   Blue         192714      5  19943
2   Honda  White          84714      4  28343
3  Toyota  White         154365      4  13434
4  Nissan   Blue         181577      3  14043

In [9]:
len(cse)

1000

In [10]:
cse.dtypes

Make             object
Colour           object
Odometer (KM)     int64
Doors             int64
Price             int64
dtype: object

In [11]:
# prepare the feature and label
x = cse.drop("Price", axis=1)
y = cse.Price

In [12]:
# Splitting the data into test and training sets (Practice time!)
from sklearn.model_selection import train_test_split # No need to do it again, it's already done above. But still... :P
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.2) 

In [13]:
x_train.size, x_test.size, y_train.size, y_test.size

(3200, 800, 800, 200)

In [14]:
# build ML model
from sklearn.ensemble import RandomForestRegressor

Now this random forest reggressor is the same as a classifier random forest. <br>
But this time it can predict a number which is what we're trying to do. <br>
We're trying to predict the price of a car given we are given some attributes about it. <br>

#### Now, our ML algorithm cannot deal with strings! So we need to take care of that --

In [15]:
# Converting categorical values (strings) into numerical values
from sklearn.preprocessing import OneHotEncoder 
from sklearn.compose import ColumnTransformer



In [16]:
model = RandomForestRegressor() #create a model
model.fit(x_train, y_train) #train a model
model.score(x_test, y_train) #run/evaluate a model

ValueError: could not convert string to float: 'Toyota'